src/retry.rs

```diff
@@ -0,0 +1,450 @@
+            assert_eq!(iter.next(), Some(Duration::from_millis(10)));
+            assert_eq!(iter.next(), Some(Duration::from_millis(10)));
+            assert_eq!(iter.next(), Some(Duration::from_millis(20)));
+            assert_eq!(iter.next(), Some(Duration::from_millis(30)));
+            assert_eq!(iter.next(), Some(Duration::from_millis(50)));
+            assert_eq!(iter.next(), Some(Duration::from_millis(50)));
+        }
+
+        #[test]
+        fn test_fibonacci_returns_max_when_max_less_than_base() {
+            let mut iter = FibonacciBackoff::from_secs(20).max_delay(Duration::from_secs(10));
+
+            assert_eq!(iter.next(), Some(Duration::from_secs(10)));
+            assert_eq!(iter.next(), Some(Duration::from_secs(10)));
+        }
+
+        #[test]
+        fn test_exponential_some_exponential_base_10() {
+            let mut s = ExponentialBackoff::from_millis(10);
+
+            assert_eq!(s.next(), Some(Duration::from_millis(10)));
+            assert_eq!(s.next(), Some(Duration::from_millis(100)));
+            assert_eq!(s.next(), Some(Duration::from_millis(1000)));
+        }
+
+        #[test]
+        fn test_exponential_some_exponential_base_2() {
+            let mut s = ExponentialBackoff::from_millis(2);
+
+            assert_eq!(s.next(), Some(Duration::from_millis(2)));
+            assert_eq!(s.next(), Some(Duration::from_millis(4)));
+            assert_eq!(s.next(), Some(Duration::from_millis(8)));
+        }
+
+        #[test]
+        fn test_exponential_saturates_at_maximum_value() {
+            let mut s = ExponentialBackoff::from_millis(u64::MAX - 1);
+
+            assert_eq!(s.next(), Some(Duration::from_millis(u64::MAX - 1)));
+            assert_eq!(s.next(), Some(Duration::from_millis(u64::MAX)));
+            assert_eq!(s.next(), Some(Duration::from_millis(u64::MAX)));
+        }
+
+        #[test]
+        fn test_exponential_stops_increasing_at_max_delay() {
+            let mut s = ExponentialBackoff::from_millis(2).max_delay(Duration::from_millis(4));
+
+            assert_eq!(s.next(), Some(Duration::from_millis(2)));
+            assert_eq!(s.next(), Some(Duration::from_millis(4)));
+            assert_eq!(s.next(), Some(Duration::from_millis(4)));
+        }
+
+        #[test]
+        fn test_exponential_max_when_max_less_than_base() {
+            let mut s = ExponentialBackoff::from_millis(20).max_delay(Duration::from_millis(10));
+
+            assert_eq!(s.next(), Some(Duration::from_millis(10)));
+            assert_eq!(s.next(), Some(Duration::from_millis(10)));
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::io::ErrorKind;
+    use std::ops::AddAssign;
+    use std::time::Duration;
+    use tracing::debug;
+
+    #[fluvio_future::test]
+    async fn test_fixed_retries_no_delay() {
+        let mut executed_retries = 0u8;
+        let operation = || {
```

Comment:
how is compared with https://docs.rs/fure/0.6.0/fure/?

Reply:
It solves the same problem but our API is neater, in my opinion, and easier to extend if needed.  We use our primitive `sleep` which has wasm32 target support and runtime-agnostic, whereas in that crate we will need additional efforts to be able to compile. I think there is not much sense to add this dependency when our own impl  is couple hundred lines and the crate is not actively developed and popular.

Reply:
fair enough